fdbserver/LeaderElection.actor.cpp

```diff
@@ -25,28 +25,56 @@
 #include "fdbclient/MonitorLeader.h"
 #include "flow/actorcompiler.h" // This must be the last #include.
 
+// Keep trying to become a leader by submitting itself to all coordinators.
+// Monitor the health of all coordinators at the same time.
+// Note: for coordinators whose NetworkAddress is parsed out of a hostname, a connection failure will cause this actor
+// to throw `coordinators_changed()` error
 ACTOR Future<Void> submitCandidacy(Key key,
                                    LeaderElectionRegInterface coord,
                                    LeaderInfo myInfo,
                                    UID prevChangeID,
-                                   Reference<AsyncVar<std::vector<Optional<LeaderInfo>>>> nominees,
-                                   int index) {
+                                   AsyncTrigger* nomineeChange,
+                                   Optional<LeaderInfo>* nominee,
+                                   Optional<Hostname> hostname = Optional<Hostname>()) {
 	loop {
-		auto const& nom = nominees->get()[index];
-		Optional<LeaderInfo> li = wait(
-		    retryBrokenPromise(coord.candidacy,
-		                       CandidacyRequest(key, myInfo, nom.present() ? nom.get().changeID : UID(), prevChangeID),
-		                       TaskPriority::CoordinationReply));
+		state Optional<LeaderInfo> li;
 
-		if (li != nominees->get()[index]) {
-			std::vector<Optional<LeaderInfo>> v = nominees->get();
-			v[index] = li;
-			nominees->set(v);
+		if (coord.candidacy.getEndpoint().getPrimaryAddress().fromHostname) {
+			state ErrorOr<Optional<LeaderInfo>> rep = wait(coord.candidacy.tryGetReply(
+			    CandidacyRequest(key, myInfo, nominee->present() ? nominee->get().changeID : UID(), prevChangeID),
+			    TaskPriority::CoordinationReply));
+			if (rep.isError()) {
+				// Connecting to nominee failed, most likely due to connection failed.
+				TraceEvent("SubmitCandadicyError")
+				    .error(rep.getError())
+				    .detail("Hostname", hostname.present() ? hostname.get().toString() : "UnknownHostname")
+				    .detail("OldAddr", coord.candidacy.getEndpoint().getPrimaryAddress().toString());
+				if (rep.getError().code() == error_code_request_maybe_delivered) {
+					// 50 milliseconds delay to prevent tight resolving loop due to outdated DNS cache
+					wait(delay(0.05));
```

Comment:
Can you use a knob for this value? Also, I feel the wait is better to be added at the caller (around line 236).

Reply:
> How did you find this issue?

John found it and reported in #6593.

Reply:
Can you change the delay to a knob? In case the looping is too aggressive, this gives us the ability make it slower on live.